backend/v3/domain/session_factor.go

```diff
@@ -0,0 +1,250 @@
+			return f, true
+		}
+	}
+	return nilT, false
+}
+
+type SessionChallenge interface {
+	sessionChallengeType() SessionFactorType
+}
+
+type SessionChallengePasskey struct {
+	LastChallengedAt     time.Time
+	Challenge            string
+	AllowedCredentialIDs [][]byte
+	UserVerification     domain.UserVerificationRequirement
+	RPID                 string
+}
+
+// sessionChallengeType implements [SessionChallenge].
+func (s *SessionChallengePasskey) sessionChallengeType() SessionFactorType {
+	return SessionFactorTypePasskey
+}
+
+type SessionChallengeOTPSMS struct {
+	LastChallengedAt  time.Time
+	Code              *crypto.CryptoValue
+	Expiry            time.Duration
+	CodeReturned      bool
+	GeneratorID       string
+	TriggeredAtOrigin string
+}
+
+// sessionChallengeType implements [SessionChallenge].
+func (s *SessionChallengeOTPSMS) sessionChallengeType() SessionFactorType {
+	return SessionFactorTypeOTPSMS
+}
+
+type SessionChallengeOTPEmail struct {
+	LastChallengedAt  time.Time
+	Code              *crypto.CryptoValue
+	Expiry            time.Duration
+	CodeReturned      bool
+	URLTmpl           string
+	TriggeredAtOrigin string
+}
+
+// sessionChallengeType implements [SessionChallenge].
+func (s *SessionChallengeOTPEmail) sessionChallengeType() SessionFactorType {
+	return SessionFactorTypeOTPEmail
+}
+
+type SessionChallenges []SessionChallenge
+
+// GetPasskeyChallenge returns the passkey challenge from the challenges.
+func (s SessionChallenges) GetPasskeyChallenge() *SessionChallengePasskey {
+	challenge, _ := GetChallenge[*SessionChallengePasskey](s)
+	return challenge
+}
+
+// GetOTPSMSChallenge returns the OTP SMS challenge from the challenges.
+func (s SessionChallenges) GetOTPSMSChallenge() *SessionChallengeOTPSMS {
+	challenge, _ := GetChallenge[*SessionChallengeOTPSMS](s)
+	return challenge
+}
+
+// GetOTPEmailChallenge returns the OTP Email challenge from the challenges.
+func (s SessionChallenges) GetOTPEmailChallenge() *SessionChallengeOTPEmail {
+	challenge, _ := GetChallenge[*SessionChallengeOTPEmail](s)
+	return challenge
+}
+
+// GetChallenge returns the challenge of the given type from the challenges.
+func GetChallenge[T SessionChallenge](challenges SessionChallenges) (T, bool) {
+	var nilT T
+	for _, challenge := range challenges {
+		if c, ok := challenge.(T); ok {
+			return c, true
+		}
+	}
+	return nilT, false
+}
+
+type sessionFactorColumns interface {
+	// InstanceIDColumn returns the column for the instance id field.
+	InstanceIDColumn() database.Column
+	// SessionIDColumn returns the column for the session id field.
+	SessionIDColumn() database.Column
+	// FactorTypeColumn returns the column for the factor type field.
+	FactorTypeColumn() database.Column
+	// LastChallengedAtColumn returns the column for the last challenged at field.
+	LastChallengedAtColumn() database.Column
+	// LastFailedAtColumn returns the column for the last failed at field.
+	LastFailedAtColumn() database.Column
```

Comment:
I don't see this as a param in any of the factor structs'

Reply:
added them where needed